rp2040-hal/src/timer.rs

```diff
@@ -289,6 +311,51 @@ macro_rules! impl_alarm {
                 }
             }
 
+            /// Schedule the alarm to be finished at the given timestamp. If [enable_interrupt] is
+            /// called, this will trigger interrupt `
+            #[doc = $int_name]
+            /// ` whenever this timestamp is reached.
+            ///
+            /// The RP2040 has been observed to take a little while to schedule an alarm. For this
+            /// reason, this function requires the timestamp to be at least 10 microseconds in the
+            /// future.
+            ///
+            /// The RP2040 is unable to schedule an event taking place in more than
+            /// `u32::max_value()` microseconds.
+            ///
+            /// [enable_interrupt]: #method.enable_interrupt
+            fn schedule_at(&mut self, timestamp: Instant) -> Result<(), ScheduleAlarmError> {
+                let timestamp_low = (timestamp.ticks() & 0xFFFF_FFFF) as u32;
+
+                // safety: Only read operations are made on the timer and they should not have any UB
+                let timer = unsafe { &*TIMER::ptr() };
+                let now = get_counter(timer);
+                let duration = timestamp.ticks().saturating_sub(now.ticks());
+                if duration > u32::max_value().into() {
+                    return Err(ScheduleAlarmError::AlarmTooLate);
+                }
+
+                cortex_m::interrupt::free(|_| {
+                    let alarm = &timer.$timer_alarm;
+
+                    // safety: This is the only code in the codebase that accesses memory address $timer_alarm
+                    alarm.write(|w| unsafe { w.bits(timestamp_low) });
+
+                    // If it is not set, it has already triggered.
+                    if (timer.armed.read().bits() & $armed_bit_mask) != 0 {
```

Comment:
Is this line needed?
It could be seen as an optimization, as it avoids the call to `get_counter` in case the alarm already triggered. However, this seems rather unlikely, as it was only scheduled a few cycles earlier.

I think the check below is sufficient. If `now < timestamp`, we are sure that the alarm was scheduled in the future and everything is fine. If `now < timestamp`, it _might_ have been scheduled in the past (and will trigger in ~72 minutes), or it _might_ have been scheduled properly and already triggered in the mean time. Those two cases are properly separated by the `(timer.armed.read().bits() & $armed_bit_mask) != 0` comparison below.

Reply:
Indeed, thanks!